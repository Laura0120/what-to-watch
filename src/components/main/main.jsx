import React from 'react';
import {connect} from "react-redux";

import {changeFavorite, fetchFavorite} from '../../store/api-actions';
import {getMoviesByGenre, getActivGenre, getGenreList} from '../../store/selectors';
import {ActionCreator} from '../../store/action';
import {FUNCTION, MOVIE, MOVIES, GENRE_LIST, STRING} from '../../prop-type';
import GenreList from '../genre-list/genre-list';
import MovieListWrapped from '../movie-list/movie-list';
import UserBlock from '../user-block/user-block';
import MovieButtons from '../movie-buttons/movie-buttons';
import {AppRoute} from '../../const';
import {adaptToClientMovie} from '../../utils/adapt';

const Main = (props) => {
  const {genreList, currentMovies, promoMovie, activeGenre, onChangeGenre, onMovieClick, onFavoriteClick, onPlayClick, authorizationStatus, onMyListButtonClick} = props;
  const {poster, backgroundImage, title, genre, year, id, isFavorite} = promoMovie || {};

  return (
    <React.Fragment>
      <section className='movie-card'>
        <div className='movie-card__bg'>
          <img src={backgroundImage} alt='The Grand Budapest Hotel' />
        </div>

        <h1 className='visually-hidden'>WTW</h1>

        <header className='page-header movie-card__head'>
          <div className='logo'>
            <a className='logo__link'>
              <span className='logo__letter logo__letter--1'>W</span>
              <span className='logo__letter logo__letter--2'>T</span>
              <span className='logo__letter logo__letter--3'>W</span>
            </a>
          </div>

          <UserBlock authorizationStatus={authorizationStatus} onMyListButtonClick={onMyListButtonClick}/>

        </header>

        <div className='movie-card__wrap'>
          <div className='movie-card__info'>
            <div className='movie-card__poster'>
              <img src={poster} alt={title} width='218' height='327' />
            </div>

            <div className='movie-card__desc'>
              <h2 className='movie-card__title'>{title}</h2>
              <p className='movie-card__meta'>
                <span className='movie-card__genre'>{genre}</span>
                <span className='movie-card__year'>{year}</span>
              </p>

              <MovieButtons id={id} isFavorite={isFavorite} onPlayClick={onPlayClick} onFavoriteClick={onFavoriteClick}/>

            </div>
          </div>
        </div>
      </section>

      <div className='page-content'>
        <section className='catalog'>
          <h2 className='catalog__title visually-hidden'>Catalog</h2>

          <GenreList genreList={genreList} activeGenre={activeGenre} onChangeGenre={onChangeGenre} />

          <MovieListWrapped movies={currentMovies} onMovieClick={onMovieClick}/>

        </section>

        <footer className='page-footer'>
          <div className='logo'>
            <a className='logo__link logo__link--light'>
              <span className='logo__letter logo__letter--1'>W</span>
              <span className='logo__letter logo__letter--2'>T</span>
              <span className='logo__letter logo__letter--3'>W</span>
            </a>
          </div>

          <div className='copyright'>
            <p>Â© 2019 What to watch Ltd.</p>
          </div>
        </footer>
      </div>
    </React.Fragment>
  );
};

Main.propTypes = {
  currentMovies: MOVIES.isRequired,
  genreList: GENRE_LIST,
  promoMovie: MOVIE.isRequired,
  authorizationStatus: STRING,
  activeGenre: STRING,
  onChangeGenre: FUNCTION,
  onMovieClick: FUNCTION,
  onFavoriteClick: FUNCTION,
  onPlayClick: FUNCTION,
  onMyListButtonClick: FUNCTION,
};

const mapStateToProps = (state) => ({
  genreList: getGenreList(state),
  currentMovies: getMoviesByGenre(state),
  activeGenre: getActivGenre(state),
  promoMovie: adaptToClientMovie(state.DATA.promoMovie),
  authorizationStatus: state.USER.authorizationStatus,
});

const mapDispatchToProps = (dispatch) => ({
  onMovieClick(id) {
    dispatch(ActionCreator.redirectToRoute(`/films/${id}`));
  },
  onChangeGenre(activeGenre) {
    dispatch(ActionCreator.changeGenre(activeGenre.target.textContent));
  },
  onFavoriteClick(id, isFavorite) {
    dispatch(changeFavorite(id, isFavorite));
  },
  onPlayClick(id) {
    dispatch(ActionCreator.redirectToRoute(`/player/${id}`));
  },
  onMyListButtonClick() {
    dispatch(fetchFavorite());
    dispatch(ActionCreator.redirectToRoute(AppRoute.MY_LIST));
  },
});

export {Main};
export default connect(mapStateToProps, mapDispatchToProps)(Main);
