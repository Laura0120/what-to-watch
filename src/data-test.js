export const movie = {
  id: 1,
  poster: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Macbeth.jpg`,
  preview: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/macbeth.jpg`,
  backgroundImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Macbeth.jpg`,
  backgroundColor: `#F1E9CE`,
  video: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
  previewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  title: `Macbeth`,
  rating: {
    ratingScore: 3.3,
    countVotesRating: 48798
  },
  director: `Justin Kurzel`,
  starring: [`Michael Fassbender`, `Marion Cotillard`, `Jack Madigan`],
  year: 2015,
  runtime: 113,
  genre: `Drama`,
  description: `Macbeth, the Thane of Glamis, receives a prophecy from a trio of witches that one day he will become King of Scotland. Consumed by ambition and spurred to action by his wife, Macbeth murders his king and takes the throne for himself.`,
  isFavorite: false
};


export const noop = () => {};

export const comment = {
  comment: `A movie that will take you to another world full of emotions.`,
  date: `2020-11-03T13:38:44.769Z`,
  id: 1,
  rating: 8.6
};
