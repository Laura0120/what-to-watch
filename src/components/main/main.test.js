import React from 'react';
import renderer from 'react-test-renderer';

import {Main} from './main';
import {movie, noop} from '../../data-test';

it(`Main is rendered correctly`, () => {
  const tree = renderer.create(
      <Main
        genreList={[movie.genre]}
        currentMovies={[movie]}
        promoMovie={movie}
        activeGenre={`All genres`}
        authorizationStatus = {`AUTH`}
        onChangeGenre={noop}
        onMovieClick={noop}
        onFavoriteClick={noop}
        onPlayClick={noop}
        onMyListButtonClick={noop}
      />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
