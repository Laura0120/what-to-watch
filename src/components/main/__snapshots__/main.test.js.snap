// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main is rendered correctly 1`] = `
Array [
  <section
    className="movie-card"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="The Grand Budapest Hotel"
        src="https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Macbeth.jpg"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header movie-card__head"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="user-block"
      >
        <div
          className="user-block__avatar"
          onClick={[Function]}
        >
          <img
            alt="User avatar"
            height="63"
            src="img/avatar.jpg"
            width="63"
          />
        </div>
      </div>
    </header>
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster"
        >
          <img
            alt="Macbeth"
            height="327"
            src="https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Macbeth.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            Macbeth
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Drama
            </span>
            <span
              className="movie-card__year"
            >
              2015
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              className="btn btn--list movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlinkHref="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
            
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <ul
        className="catalog__genres-list"
      >
        <li
          className="catalog__genres-item "
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Drama
          </a>
        </li>
      </ul>
      <div
        className="catalog__movies-list"
      >
        <article
          className="small-movie-card catalog__movies-card"
          id={1}
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <video
              height="175"
              muted={true}
              poster="https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/macbeth.jpg"
              src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="#"
            >
              Macbeth
            </a>
          </h3>
        </article>
      </div>
      
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
